prow/config/cmd/generate.go

```diff
@@ -62,50 +63,61 @@ func main() {
 		panic("too many arguments")
 	}
 
-	var settings config.GlobalConfig
-	if _, err := os.Stat(filepath.Join(*inputDir, ".global.yaml")); !os.IsNotExist(err) {
-		settings = config.ReadGlobalSettings(filepath.Join(*inputDir, ".global.yaml"))
+	var bc *config.BaseConfig
+	if _, err := os.Stat(filepath.Join(*inputDir, ".base.yaml")); !os.IsNotExist(err) {
+		bc = config.ReadBase(nil, filepath.Join(*inputDir, ".base.yaml"))
 	}
-	cli := &config.Client{GlobalConfig: settings}
 
 	if os.Args[1] == "branch" {
-		if err := filepath.Walk(*inputDir, func(src string, file os.FileInfo, err error) error {
+		if err := filepath.WalkDir(*inputDir, func(path string, d os.DirEntry, err error) error {
 			if err != nil {
 				fmt.Printf("error: %s\n", err.Error())
 			}
-
-			if file.IsDir() {
-				return nil
-			}
-			if filepath.Ext(file.Name()) != ".yaml" && filepath.Ext(file.Name()) != ".yml" || file.Name() == ".global.yaml" {
-				log.Println("skipping", file.Name())
-				return nil
+			baseConfig := bc
+			if _, err := os.Stat(filepath.Join(path, ".base.yaml")); !os.IsNotExist(err) {
+				baseConfig = config.ReadBase(baseConfig, filepath.Join(path, ".base.yaml"))
 			}
-			jobs := cli.ReadJobsConfig(src)
-			jobs.Jobs = config.FilterReleaseBranchingJobs(jobs.Jobs)
-
-			if jobs.SupportReleaseBranching {
-				tagRegex := regexp.MustCompile(`^(.+):(.+)-([0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}-[0-9]{2}-[0-9]{2})$`)
-				match := tagRegex.FindStringSubmatch(jobs.Image)
-				branch := "release-" + flag.Arg(1)
-				if len(match) == 4 {
-					newImage := fmt.Sprintf("%s:%s-%s", match[1], branch, match[3])
-					if err := exec.Command("gcloud", "container", "images", "add-tag", match[0], newImage).Run(); err != nil {
-						exit(err, "unable to add image tag: "+newImage)
-					} else {
-						jobs.Image = newImage
-					}
+			cli := config.Client{BaseConfig: baseConfig}
+
+			files, _ := ioutil.ReadDir(path)
+			for _, file := range files {
+				if file.IsDir() {
+					continue
+				}
+
+				if (filepath.Ext(file.Name()) != ".yaml" && filepath.Ext(file.Name()) != ".yml") ||
+					file.Name() == ".base.yaml" {
+					log.Println("skipping", file.Name())
+					continue
 				}
-				jobs.Branches = []string{branch}
-				jobs.SupportReleaseBranching = false
 
-				name := file.Name()
-				ext := filepath.Ext(name)
-				name = name[:len(name)-len(ext)] + "-" + flag.Arg(1) + ext
+				src := filepath.Join(path, file.Name())
+				jobs := cli.ReadJobsConfig(src)
+				jobs.Jobs = config.FilterReleaseBranchingJobs(jobs.Jobs)
+
+				if jobs.SupportReleaseBranching {
+					tagRegex := regexp.MustCompile(`^(.+):(.+)-([0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}-[0-9]{2}-[0-9]{2})$`)
+					match := tagRegex.FindStringSubmatch(jobs.Image)
+					branch := "release-" + flag.Arg(1)
+					if len(match) == 4 {
```

Comment:
Can you add some docs to clarify the logic here a bit? It also might be useful to have a utility break some of this down into clear method names that aid in the reader's understanding.

Reply:
This change isn't quite related to my change (mostly due to other refactoring/renaming), but I've added some comments here based on my understanding. It looks some improvements can also be done here in the future, e.g. reading the latest tag for the image of the new release branch, instead of adding a new tag for the image of the current branch.